Implement synchronization scheme for incr. comp. directory

michaelwoerister opened the pull request:
This PR implements a copy-on-write-based synchronization scheme for the incremental compilation cache directory. For technical details, see the documentation at the beginning of `rustc_incremental/persist/fs.rs`.

The PR contains unit tests for some functions but for testing whether the scheme properly handles races, a more elaborate test setup would be needed. It would probably involve a small tool that allows to manipulate the incremental compilation directory in a controlled way and then letting a compiler instance run against directories in different states. I don't know if it's worth the trouble of adding another test category to `compiletest`, but I'd be happy to do so.

Fixes #32754
Fixes #34957

rust-highfive commented:
r? @nrc

(rust_highfive has picked a reviewer for you, use r? to override)

michaelwoerister commented:
cc @nikomatsakis @alexcrichton @brson @rust-lang/compiler

Mark-Simulacrum commented on src/librustc/session/mod.rs:
```diff
@@ -331,6 +334,105 @@ impl Session {
+                bug!("Trying initialize already active IncrCompSession.")
+            }
+            IncrCompSession::Finalized { .. } => {
+                bug!("Trying initialize already finalized IncrCompSession.")
```
nit: Both this bug! and the one before it should be "Trying to initialize.."

michaelwoerister replied:
thanks!

Mark-Simulacrum commented on src/librustc/util/fs.rs:
```diff
@@ -56,14 +56,49 @@ pub fn fix_windows_verbatim_for_gcc(p: &Path) -> PathBuf {
 /// Copy `p` into `q`, preferring to use hard-linking if possible. If
 /// `q` already exists, it is removed first.
-pub fn link_or_copy<P: AsRef<Path>, Q: AsRef<Path>>(p: P, q: Q) -> io::Result<()> {
+/// The result indicates which of the to operations has been performed.
```
nit: `of the two operations` instead of `of the to operations`.

Mark-Simulacrum commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,810 @@
+//! synchronization will go haywire.
+//! In both cases we recommend to locate the incremental compilation directory
+//! on a file system that supports these things.
+//! I might be a good idea though to try and detect whether we are on an
```
nit: `It might`

Mark-Simulacrum commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,810 @@
+    let mut deletion_candidates = vec![];
+    let mut definitely_delete = vec![];
+
+    for dir_entry in try!(crate_directory.read_dir()) {
```
Skipping files found may also be a good idea to prevent odd issues later (`entry.file_type().is_dir()`, but `file_type` returns a Result).

michaelwoerister replied:
Sure, I'll add it.

Mark-Simulacrum commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,810 @@
+        match extract_timestamp_from_session_dir(&directory_name) {
+            Ok(timestamp) => {
+                if directory_name.ends_with("-working") {
+                    let ten_seconds = Duration::new(10, 0);
```
`Duration::from_secs(10)` is clearer to me.

michaelwoerister replied:
:+1:

Mark-Simulacrum commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,810 @@
+                    // chance of deleting a directory in the time window where
+                    // the process has allocated the directory but has not yet
+                    // acquired the file-lock on it.
+                    if timestamp < SystemTime::now() - ten_seconds {
```
Using `SystemTime::now().duration_since(timestamp) > ten_seconds` is nicer, but may fail (since duration_since could be "negative"). Another alternative is `timestamp + ten_seconds < SystemTime::now()`.

Mark-Simulacrum commented on src/librustc_incremental/persist/save.rs:
```diff
@@ -39,31 +39,31 @@ pub fn save_dep_graph<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>) {
-    // FIXME(#32754) lock file?
-
     // delete the old dep-graph, if any
+    // Note: It's important that we actually delete the old file and note just
```
nit: `note` -> `not`

nagisa commented on src/librustc/session/mod.rs:
```diff
@@ -331,6 +334,105 @@ impl Session {
+        let mut incr_comp_session = self.incr_comp_session.borrow_mut();
+
+        match *incr_comp_session {
+            IncrCompSession::NotInitialized => {
```
Why can’t all these be just a 

```
if let IncrCompSession::Active = *incr_comp_session { } else { 
    bug!("trying to finalize {:?} session", incr_comp_session);
}
```

The bug message really doesn’t matter that much to have so much extra code in each state transition dedicated purely for reporting.

michaelwoerister replied:
As long as I can see the current state of the session in the error message, I'm fine with that.

nagisa commented on src/librustc/util/fs.rs:
```diff
@@ -56,14 +56,49 @@ pub fn fix_windows_verbatim_for_gcc(p: &Path) -> PathBuf {
-    fs::hard_link(p, q)
-        .or_else(|_| fs::copy(p, q).map(|_| ()))
+
+    match fs::hard_link(p, q) {
```
This will be horribly slow on Windows using ReFS filesystems. The [CreateHardLink](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363860(v=vs.85).aspx) is only supported on the NTFS systems. ReFS supports copy-on-write (read: reflinks), but not hard links, so it is yet another reason to use reflinks if these are supported.

michaelwoerister replied:
I can't find any documentation about reflinks on ReFS, so I guess that any copy operation within the same ReFS file system is automatically the equivalent of creating a reflink? In that case, we could just fall back to always copying if the first try of creating a hard-link has failed.

nagisa replied:
AFAICT simple copy is not reflink on ReFS, sadly.

There’s [FSCTL](https://msdn.microsoft.com/en-us/library/windows/desktop/mt590823(v=vs.85).aspx) in a Windows Server 2016+ which seems like it should behave like the reflink equivalent for btrfs.

That being said, it also seems like ReFSv1 (which is what you get by default) does not in fact support/expose that operation, so I guess there’s nothing we can do here.

michaelwoerister replied:
Well, that's a pity `:(`
But thanks for the information!

nagisa commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,810 @@
+//!
+//! Another case that has to be considered is what happens if one process
+//! deletes a finalized session directory that another process is currently
+//! trying to copy from. In that case, depending on the exact operating system
```
I feel like this problem is trivially solved by simply locking the directory lock file while deleting its contents or trying to copy/link from it.

In case you’re deleting/garbage collecting and cannot lock the file, it means somebody else is either already doing the deletion/garbage collection or is using/copying from the directory in question.

In case you’re copying and you find a locked file, its either that somebody is deleting this directory and you should try to find a newer one, or somebody else is copying from this directory and you should wait. The wait here is suboptimal, and could be solved by taking an exclusive (write) lock for deletion and shared (read) lock for copy.

michaelwoerister replied:
RWLock idea is interesting.

michaelwoerister replied:
I'll implement this tomorrow. I'm interested if it will work on Windows (deleting a directory while holding a lock on a file in it).

michaelwoerister replied:
So, this *should* on Windows, as long as the lock file is always opened with `FILE_SHARE_DELETE` by the process that wants to delete it.

nagisa replied:
FILE_SHARE_DELETE seems to only be necessary if the lock file is intended to be shared? In case your process is the sole owner of a handle to the file (as you’re supposed to be when deleting) then it should be fine even without that flag?

nagisa replied:
That being said, if you want to such race conditions such as:

* rustc1 holds the exclusive lock and is deleting the contents of the directory;
* rustc2 tries to inspect a directory, enters it, finds it having no lock file (rustc1 just deleted the lock file before deleting the directory) and thus tries to take the lock itself.
* rustc1 cannot delete the directory because it contains the lock just created by rustc2.

You probably want to have the lock file outside the directory in the first place. i.e. something like:

|- rustc-incremental-dir
|- rustc-incremental-dir.lock

Then you can delete the directory while holding the lock without any concerns and you’d avoid the race condition described above, because even if rustc did this:

* find directory (not deleted yet);
* try to lock its respective lock (directory and lock are cleaned up already), and succeed doing so;
* try working with the directory (directory does not exist anymore, all operations fail).

michaelwoerister replied:
> FILE_SHARE_DELETE seems to only be necessary if the lock file is intended to be shared during deletion? In case your process is the sole owner of a handle to the file (as you’re supposed to be when deleting) then it should be fine even without that flag?

I've tested this with a small C program yesterday and if I didn't specify FILE_SHARE_DELETE, not even the process owning the handle could delete the file via `DeleteFile` (which is what `std::fs::remove_dir_all` uses). 

> 1. rustc1 holds the exclusive lock and is deleting the contents of the directory;
> 2. rustc2 tries to inspect a directory, enters it, finds it having no lock file (rustc1 just deleted the lock file before deleting the directory) and thus tries to take the lock itself.
> 3. rustc1 cannot delete the directory because it contains the lock just created by rustc2.

This condition would be handled because a garbage collecting process would not try to do anything further with a directory without a lock-file.

michaelwoerister replied:
But I agree that having the lock-file outside of the directory it locks seems a bit cleaner.

nrc commented:
r? @nikomatsakis

michaelwoerister commented:
OK, I've implemented the read-write-lock approach. That seems like a clear improvement. I did not, however, move the lock file out of the directory. It made cleanup code more complicated in a few places and having it *in* the directory should be just as correct. I think, I also addressed all the nits.

nikomatsakis commented on src/librustc/util/fs.rs:
```diff
@@ -56,14 +56,49 @@ pub fn fix_windows_verbatim_for_gcc(p: &Path) -> PathBuf {
+
+// Like std::fs::create_dir_all, except handles concurrent calls among multiple
+// threads or processes.
+pub fn create_dir_racy(path: &Path) -> io::Result<()> {
```
is this just copied from compiletest?

michaelwoerister replied:
It's moved here from `rustc_incremental::persist::util` which doesn't exist anymore.

nikomatsakis commented on src/librustc_data_structures/flock.rs:
```diff
@@ -214,24 +252,72 @@ mod imp {
+                   exclusive: bool)
+                   -> io::Result<Lock> {
+
+            let share_mode = FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE;
```
gonna trust you on all this grody code :)

nikomatsakis commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,862 @@
+
+
+//! This module manages how the incremental compilation cache is represented in
+//! the file system.
```
❤️ your comments

nikomatsakis commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,862 @@
+        // {incr-comp-dir}/{crate-name-and-disambiguator}/sess-{timestamp}-{random}-working
+        //
+        // If this fails, return an error, don't retry
+        let session_dir = try!(alloc_session_dir(tcx.sess, &crate_dir));
```
Nit: you can use `?` now :)

michaelwoerister replied:
I actually find `try!` more readable. The `?` is so easily overlooked.

nikomatsakis commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,862 @@
+        // Drop the file lock
+        sess.mark_incr_comp_session_as_invalid();
+        if let Err(err) = std_fs::remove_dir_all(&*incr_comp_session_dir) {
+            debug!("finalize_session_directory: Failed to delete invalidated \
```
maybe issue a warning? use might care.

michaelwoerister replied:
You think so? Garbage collection will take care of it eventually.

nikomatsakis commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,862 @@
+    where I: Iterator<Item=String>
+{
+    for sub_dir_name in iter {
+        if sub_dir_name.ends_with("-working") || !sub_dir_name.starts_with("sess-") {
```
might be nice to encapsulate this into a helper, not sure how often it's repeated

nikomatsakis commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -0,0 +1,862 @@
+
+/// Finds the correct metadata hashes file for the given crate. In order to do
+/// that it has to find the crate directory of the given crate, and in there,
+/// look for the session directory with the correct SVH in it.
```
Do we not want some sort of read file-lock here? I worry that we will be in the middle of reading from a version of a crate when it is being recompiled. Are we not trying to defend against that scenario? Seems like it'd cause problems in other ways anyhow.

nikomatsakis replied:
Probably worth spelling out (in the mega comment, I imagine) just what threat scenarios we are and are not defending against here!

nikomatsakis commented:
ok, r=me modulo nits

michaelwoerister commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit d1d8258 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis d1d8258c5461e26f54c4ec1b75c5cf7779478141 -->

bors commented:
:hourglass: Testing commit d1d8258 with merge 9f826a9...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2216)

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 95d3cf3 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 95d3cf3353111e5e19820fa1eae58060de58f47f -->

bors commented:
:hourglass: Testing commit 95d3cf3 with merge da5ce20...

bors commented:
:broken_heart: Test failed - [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3419)

alexcrichton commented:
@bors: retry

On Tue, Aug 23, 2016 at 11:37 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - auto-linux-cross-opt
> <https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3419>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/35718#issuecomment-241831943>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95OOMbfqzXTBCXRyrGTpkKz1jxiu4ks5qiz3ogaJpZM4Jljxv>
> .
>

bors commented:
:umbrella: The latest upstream changes (presumably #35854) made this pull request unmergeable. Please resolve the merge conflicts.

michaelwoerister commented:
@bors r=nikomatsakis

rebased

bors commented:
:pushpin: Commit 33bd082 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 33bd082cf288cec0f677d1d10315253558b511f9 -->

bors commented:
:hourglass: Testing commit 33bd082 with merge 7947f4b...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2256)

michaelwoerister commented:
@bors r=nikomatsakis

accidentally pushed an outdated branch yesterday.

bors commented:
:pushpin: Commit 7c0ad44 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 7c0ad4422471134ca7e76e1416ea0a7e3038ab7d -->

bors commented:
:hourglass: Testing commit 7c0ad44 with merge 39743d8...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1593)

michaelwoerister commented:
@bors r=nikomatsakis

Fixed rustbuild

bors commented:
:pushpin: Commit 1c2a222 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 1c2a222f82938126852dd483539777000d506bac -->

bors commented:
:hourglass: Testing commit 1c2a222 with merge 4baf862...

bors commented:
:broken_heart: Test failed - [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10322)

michaelwoerister commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit e3c8b79 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis e3c8b79bb4f17ca65674f09751177e44318924c9 -->

bors commented:
:hourglass: Testing commit e3c8b79 with merge 51d8f73...

bors commented:
:snowman: The build was interrupted to prioritize another pull request.

michaelwoerister commented:
@bors r-

Since the long session directory names are very likely to cause trouble, I want to look into making them shorter...

alexcrichton commented:
@michaelwoerister so to clarify what happened with #36017, I believe we had the following sequence of events:

1. This PR went to bors
2. This PR got all the way into the incremental tests on the two Windows rustbuild bots
3. This PR created some session directories which ended up having a too-long path name if you looked at the absolute path name.
4. This PR the bounced (presumably for some other reason)
5. The build directory was the not automatically cleaned out for the next PR (this is intentional in buildbot)
6. Future runs of the incremental tests presumably have some version of `fs::remove_dir_all` near the beginning, and then this failed because the path name was too long.
7. That panicked, causing all future builds running against the dirty directory to fail.

So I believe the only issue here was that on rustbuild the paths happened to go over the global path limit. This may affect non-rustbuild as well, but the directories are a bit longer there:

The offending pathname I saw was:

```
/c/bot/slave/auto-win-msvc-64-opt-rustbuild/build/obj/build/x86_64-pc-windows-msvc/test/incremental/ich_method_call_trait_scope.incremental/ich_method_call_trait_scope-224534f2bd4fd5de/sess-53af0b4562f62-20bf3884-9ace1227172865df/cgu-ich_method_call_trait_scope-mod3.o
```

Now this is a pretty absurd path name:

* To get to the *build directory*, it's 54 bytes (`/c/bot/slave/auto-win-msvc-64-opt-rustbuild/build/obj/`
* To get to the *test directory*, it's another 85 bytes (`build/x86_64-pc-windows-msvc/test/incremental`)
* The test itself then generates a pretty long path name of 128 bytes (`ich_method_call_trait_scope-224534f2bd4fd5de/sess-53af0b4562f62-20bf3884-9ace1227172865df/cgu-ich_method_call_trait_scope-mod3.o`)

We should probably optimize all three of these all at once. To start off with though perhaps the test could generate a shorter path? The others will be a bit more difficult to shorten I think, but can certainly be done!

eddyb commented:
@alexcrichton Can't we use `\\?\` to make the long path work?

Even ignoring the bots, AFAICT *users* will hit this, on some setups probably more often than we do on the bots (givem the large hashes in the file name).
If we can make long paths work, we should do it IMO.

alexcrichton commented:
@eddyb yes I believe that would work, as those style paths don't have the same restrictions on path limits. 

To be clear this is a problem with *compiletest* right now, not the compiler. In general though, yes, I agree that this can be an issue for users as well if the file names are too long.

michaelwoerister commented:
@alexcrichton I've pushed a few changes to that effect.

alexcrichton commented:
Looks great to me!

michaelwoerister commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 3caca7e has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 3caca7e58e2942efdd722956c9fab47e401dd13a -->

michaelwoerister commented:
@alexcrichton @eddyb This seems to be a restriction purely in the Windows API (i.e. `DeleteFileW`) and not in the underlying file system. Would it be too much of a performance hit to always prepend that magic prefix?

eddyb commented:
@michaelwoerister You could do it conditionally, since you already know the length.

michaelwoerister commented:
@eddyb yep.

Manishearth commented:
@bors r-

Tidy fails

```
/build/src/librustc_incremental/persist/fs.rs:415: trailing whitespace

/build/src/librustc_incremental/persist/fs.rs:1008: trailing whitespace

thread 'main' panicked at 'some tidy checks failed', /build/src/tools/tidy/src/main.rs:53

```

bors commented:
:umbrella: The latest upstream changes (presumably #35984) made this pull request unmergeable. Please resolve the merge conflicts.

michaelwoerister commented:
@alexcrichton Care to take a look at those last two commits and r+ if they look good to you?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit bcd2f90 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton bcd2f905c46158f9137fa5b63aafebcb60083385 -->

bors commented:
:hourglass: Testing commit bcd2f90 with merge 5a8ca4b...

alexcrichton commented:
@bors: retry force clean

* restarted buildbot

bors commented:
:hourglass: Testing commit bcd2f90 with merge 2c01bb8...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10376), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10430), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1664), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1489), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3648), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10370), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10331), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/2303), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/10308), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3530), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5480), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10409), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10382), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2327), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1506), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5496), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2357), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5472), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4664), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1682), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5477), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2353)
Approved by: alexcrichton
Pushing 2c01bb885108c436adae2006632ff6dfc0a5f2cd to master...
